### PR TITLE
Fix bug in compatibility in using old API.

### PR DESCRIPTION
**Pull request by rocky**
Given the _threat_ of a demo and the fact that over the weekend we saw that people are still using the old API_KEY, and the desire to be able to get people up to speed quickly (on prod), I see the wisdom of better compatibility with the old API_KEY. For now. 

The meat of this PR is really in one little logic bug in `index.js`.

### CODE REVIEW

**Comment by fgimenez** [comment]
I think this change would break the checks for JWT authentication. 

In API key authentication`email` is not required, it is enough to set the `apiKey` field (this has been always this way, it hasn't changed lately). 

In JWT authentication, you can set either `email` or `ethAddress`. If one of them is defined, then you need to also set `password`, this is what the check enforces:
```javascript
if ((auth.email !== undefined || auth.ethAddress !== undefined) && auth.password === undefined) {
    throw new TypeError('Please provide a password auth option.')
}
```
If you allow `email` to be undefined in this check then you could instantiate the client just with email, and you would not be able to authenticate. Let me know if I can help to make this more clear, thanks!

**Comment by rocky** [comment]
Ok. Without that, I wasn't able to get the older auth working. @daniyarchambylov will fix up my code and put in a PR. Thanks Daniyar!

**Comment by fgimenez** [comment]
> Without that, I wasn't able to get the older auth working.

Yep, you probably were setting both `email` and `apiKey`, you only need to set `apiKey` for API key based auth.

**Comment by birdofpreyru** [comment]
Noticed the piece of code above in email updates. You really should write such things as
```js
if ((auth.email || auth.ethAddress) && !auth.password) {
    throw new TypeError('Please provide a password auth option.')
}
```
it will save you from many potential problems going forward (e.g. it works as intended if fields are set to `null` rather than undefined).

**Comment by rocky** [comment]
@daniyarchambylov would you address @birdofpreyru concern? 

@birdofpreyru I'd be interested to understand in more detail what kinds of problems this would address, or in what scenary we would find this. Thanks.

**Review by fgimenez — DISMISSED** [review]
Nice, now the check really reflects the required logic. If you could address @birdofpreyru concern that would be great :+1:

**Comment by birdofpreyru** [comment]
@rocky well, may be _problems_ is too bold to say here. But when people want to "unset" a string field in JS, they have many choices: delete the object field, assign `undefined` to it, or assign `null`, or assign empty string. You write the code I propose, it will handle all these options the same and correctly; and also the condition itself looks more compact.

**Comment by rocky** [comment]
@birdofpreyru Thanks for the clarification. I think I understand now.

**Comment by daniyarchambylov** [comment]
Since @birdofpreyru raised issue of code readability and avoiding potential bugs I pushed new changes with rewritten Client constructor

File: index.js [hunk 1]
```diff
@@ -11,27 +11,29 @@ const defaultApiVersion = 'v1'
 
 class Client {
   constructor (auth, inputApiUrl = defaultApiUrl) {
-    if (auth === undefined) {
+    if (typeof auth !== 'object' || Array.isArray(auth)) {
```
Comment (birdofpreyru) [inline index.js]: I don't get it. The equivalent of the old check is just `if (!auth)`. If we want to check that `auth` is a plain object, I would add `lodash` library as dependency and use https://lodash.com/docs/4.17.11#isPlainObject check (that will be more readable, and no need to worry that we miss some edge-cases here).

Comment (fgimenez) [reply index.js]: I agree, `if (!auth)` looks ok to me here.

File: index.js [hunk 2]
```diff
@@ -11,27 +11,29 @@ const defaultApiVersion = 'v1'
...
       throw new TypeError('Please provide auth options.')
     }
 
-    if (auth.email === undefined && auth.ethAddress === undefined && auth.apiKey === undefined) {
+    const { email, ethAddress, apiKey, password } = auth
```
Comment (fgimenez) [inline index.js:18]: Nice!

**Review by fgimenez — COMMENTED** [review]
Looks good other that the empty auth check that Sergey mentioned :+1:

File: index.js (hunk 1 above)
Comment (daniyarchambylov) [reply index.js]: Ok, then we'll just check that `auth` exists and assume that `auth` is an object.

**Comment by daniyarchambylov** [comment]
@fgimenez Done!

**Comment by daniyarchambylov** [comment]
@rocky @fgimenez @birdofpreyru - Coonflicts resolved.

**Comment by rocky** [comment]
@fgimenez @birdofpreyru what is the status of this PR?

**Comment by birdofpreyru** [comment]
I am not really following this one. I guess, you can merge it in.

**Review by fgimenez — APPROVED** [review]

**Comment by fgimenez** [comment]
Yes, I thought it was already in, sorry for not approving before.

**Pull request merged**